### PR TITLE
Implement urls

### PR DESCRIPTION
**Pull request by Sammers21**
Implement urls field in Index.yaml #38

### CODE REVIEW

**Comment by 0crat** [comment]
This pull request [#38](https://github.com/artipie/helm-adapter/issues/38) is assigned to @paulodamaso[/z](https://www.0crat.com/u/paulodamaso), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/e4c58e9c-870a-4b50-a8f0-054fdca2f323); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by Sammers21** [comment]
@paulodamaso, @g4s8, This PR is ready for review.

File: src/main/java/com/artipie/helm/HelmSlice.java [hunk 1]
```diff
@@ -46,16 +46,17 @@
      * Ctor.
      *
      * @param storage The storage.
+     * @param base The base path for urls field.
      */
-    public HelmSlice(final Storage storage) {
+    public HelmSlice(final Storage storage, final String base) {
```
Comment (g4s8) [inline src/main/java/com/artipie/helm/HelmSlice.java:51]: @Sammers21 what is the base path? I think we need more documentation for public API.

File: src/main/java/com/artipie/helm/IndexYaml.java [hunk 2]
```diff
@@ -132,27 +139,28 @@ public Completable update(final TgzArchive arch) {
...
+        final ChartYaml chart = tgz.chartYaml();
         final String version = "version";
         final Map<String, Object> entries = (Map<String, Object>) index.get("entries");
         final ArrayList<Map<String, Object>> versions = (ArrayList<Map<String, Object>>)
             entries.getOrDefault(chart.field("name"), new ArrayList<Map<String, Object>>(0));
         if (versions.stream().noneMatch(map -> map.get(version).equals(chart.field(version)))) {
             final Map<String, Object> newver = new HashMap<>();
             newver.put("created", ZonedDateTime.now().format(IndexYaml.TIME_FORMATTER));
+            final ArrayList<String> urls = new ArrayList<>(1);
+            urls.add(String.format("%s%s", this.base, tgz.name()));
```
Comment (g4s8) [inline src/main/java/com/artipie/helm/IndexYaml.java:155]: @Sammers21 this URI building is not robust enough: it's possible to pass URI without leading slash. I'd use something like `org.apache.http.client.utils.URIBuilder` or similar to construct correct URIs and validate it. To build URI path you can use `Paths.get(base, tgz.name()).normalize().toString()`

File: src/main/java/com/artipie/helm/IndexYaml.java [hunk 3]
```diff
@@ -132,27 +139,28 @@ public Completable update(final TgzArchive arch) {
...
+            newver.put("urls", ZonedDateTime.now().format(IndexYaml.TIME_FORMATTER));
```
Comment (g4s8) [inline src/main/java/com/artipie/helm/IndexYaml.java:156]: @Sammers21 it can be replaced with `Collections.singletonList()`

**Comment by 0crat** [comment]
This pull request [#38](https://github.com/artipie/helm-adapter/issues/38) is assigned to @g4s8[/z](https://www.0crat.com/u/g4s8), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/04747422-cef9-4e59-8e49-44dec690f1e2); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be no monetary reward for this job

**Comment by g4s8** [comment]
@Sammers21 please fix failing builds

**Comment by Sammers21** [comment]
> @Sammers21 please fix failing builds

Done

**Comment by 0crat** [comment]
Job was finished in 27 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Pull request merged**